### ISSUE TITLE
Types from `Token!` as enum?

### ISSUE THREAD

**Issue by Evelyn-H** [dtolnay/syn#577]
Disclaimer: This is my first time using `syn`, so I might have missed something obvious...

I ran into a little issue in my parser that could easily be solved if the various `Token`'s that are provided by `syn` (the ones that are returned by the `Token! ` macro) were part of an enum, but sadly they're not. They do however have the `Token` marker trait, but I can't match on that, which is the crux of my issue.

So, basically what I'm asking is, why aren't the `Token`'s an enum?

**Comment by dtolnay** [comment]
Our idiom for parsing benefits from treating tokens (like static_token, colon_token, eq_token, semi_token in the snippet below) as syntax tree nodes defined by a Parse impl. That lets them combine with generic parse functions like `<Option<T> as Parse>::parse` and `Punctuated::<T, P>::parse_separated_nonempty`. Parsing `Punctuated` in particular would be more complicated if the punctuation token type had to be an argument rather than a type parameter.

---

https://github.com/dtolnay/syn/blob/d9d5b25b9dbf79a893b18051c94392cc14e47863/src/item.rs#L1040-L1055
